hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -864,33 +879,19 @@ protected void unblockWhenComplete(@Nonnull final BiConsumer<? super V, ? super
         }
     }
 
-    @Override
-    public InternalCompletableFuture<V> exceptionally(@Nonnull Function<Throwable, ? extends V> fn) {
-        requireNonNull(fn);
+    private void unblockExceptionally(@Nonnull Function<Throwable, ? extends V> fn,
+                                      InternalCompletableFuture<V> future) {
         Object result = resolve(state);
-        final InternalCompletableFuture<V> future = newCompletableFuture();
-        for (; ; ) {
-            if (result != UNRESOLVED && isDone()) {
```

Comment:
Here `result` may be not `UNRESOLVED` but another valid value of a non-complete future like `Thread`, `AbstractInvocationFuture$WhenCompleteNode` etc. The problem is that `isDone()` operates on field `state` but `result` reflects a potentially stale value -> it is possible that `result != UNRESOLVED` and `isDone()` are both true and yet `result` is not the actual completion value.